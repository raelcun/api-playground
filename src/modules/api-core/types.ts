import * as t from 'io-ts'
import Koa from 'koa'

import { RolesV } from '@lib/rbac'

interface KoaRequest<T> extends Koa.Request {
  body: T
}

export interface KoaContext<T> extends Koa.Context {
  request: KoaRequest<T>
}
export type KoaContext2<T> = Koa.ParameterizedContext<Koa.DefaultState, { request: { body: T } }>

type a = Koa.Middleware<Koa.DefaultState, { request: { body: T } }>

export type Middleware<T> = (ctx: KoaContext<T>, next: () => Promise<void>) => Promise<void>

export const tokenV = t.type({
  userId: t.string,
  role: RolesV,
})
export type Token = t.TypeOf<typeof tokenV>
