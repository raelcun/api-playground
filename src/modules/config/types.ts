import * as t from 'io-ts'

import { LoggingConfig } from '@lib/logger'
import { DeepPartial, NonEmptyString } from '@lib/utils'

export interface FullConfig {
  isProduction: boolean
  env: string
  server: {
    jwtSecret: string
    securePort: number
    insecurePort: number
    enableSSL: boolean
  }
  logging: LoggingConfig
  aws: {
    region: string
    credentials: {
      accessKeyId: string
      secretAccessKey: string
    }
  }
}

export type PartialConfig<T> = DeepPartial<T>

export type ConfigProvider = () => FullConfig

export const envVarsV = t.type({
  NODE_ENV: NonEmptyString,
  JWT_SECRET: NonEmptyString,
  AWS_ACCESS_KEY_ID: NonEmptyString,
  AWS_SECRET_ACCESS_KEY: NonEmptyString,
})
export type EnvVars = t.TypeOf<typeof envVarsV>
