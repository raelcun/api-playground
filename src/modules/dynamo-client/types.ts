import { DocumentClient } from 'aws-sdk/clients/dynamodb'
import { AWSError } from 'aws-sdk/lib/error'
import { Response } from 'aws-sdk/lib/response'
import { taskEither as TE } from 'fp-ts'

import { Err } from '@modules/error/types'
import { ChangeTypeOfKeys } from '@root/utils/types'

export interface DynamoClient<T> {
  put: (
    params: ChangeTypeOfKeys<DocumentClient.PutItemInput, 'Item', T>,
  ) => TE.TaskEither<Err, Response<DocumentClient.PutItemOutput, AWSError>>
  delete: (
    params: DocumentClient.DeleteItemInput,
  ) => TE.TaskEither<Err, Response<DocumentClient.DeleteItemOutput, AWSError>>
}
