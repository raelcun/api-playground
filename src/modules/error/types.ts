export interface Err<Code extends string = string> {
  code: Code
  message?: string
}
