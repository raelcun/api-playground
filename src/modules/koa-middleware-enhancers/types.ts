import Koa from 'koa'

interface KoaRequest<T> extends Koa.Request {
  body: T
}

export interface KoaContext<T> extends Koa.Context {
  request: KoaRequest<T>
}
