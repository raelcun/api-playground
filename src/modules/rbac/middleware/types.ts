import * as t from 'io-ts'

import { RolesV } from '../types'

export const tokenV = t.type({
  userId: t.string,
  role: RolesV,
})
export type Token = t.TypeOf<typeof tokenV>
