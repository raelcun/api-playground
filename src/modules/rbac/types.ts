import * as t from 'io-ts'

export const SubjectV = t.union([t.literal('user'), t.literal('admin')])
export type Subject = t.TypeOf<typeof SubjectV>

export interface Actions {
  account: 'viewAny' | 'viewOwn' | 'editOwn' | 'editAny'
}

export type Enforce = <T extends keyof Actions, U extends Actions[T]>(
  subject: Subject,
  resource: T,
  ...actions: U[]
) => Promise<boolean>
