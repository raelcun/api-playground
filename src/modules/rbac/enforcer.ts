import { newModel, Enforcer } from 'casbin'
import * as TE from 'fp-ts/lib/TaskEither'
import * as E from 'fp-ts/lib/Either'
import { pipe } from 'fp-ts/lib/pipeable'
import { createVoidTE } from 'utils'
import { Err } from '../error/types'
import { Roles, Enforce, Actions, EnforceProvider } from './types'
import { newEnforcer, enforce, addPoliciesToEnforcer } from './wrapper'

const setEnforcerModel = (enforcer: Enforcer): TE.TaskEither<Err, Enforcer> =>
  TE.tryCatch(
    async () => {
      enforcer.setModel(
        newModel(`
          [request_definition]
          r = sub, obj, act

          [policy_definition]
          p = sub, obj, act

          [role_definition]
          g = _, _

          [policy_effect]
          e = some(where (p.eft == allow))

          [matchers]
          m = g(r.sub, p.sub) && r.obj == p.obj && r.act == p.act || r.sub == "admin"
        `),
      )
      return enforcer
    },
    () => ({ code: 'ENFORCER_SETMODEL_FAILED', message: 'failed to set enforcer model' }),
  )

export const createEnforcer: TE.TaskEither<Err, Enforcer> = pipe(
  newEnforcer,
  TE.chain(setEnforcerModel),
  TE.chain(addPoliciesToEnforcer(Roles.User, 'account', 'viewOwn', 'viewAny')),
)

let enforcerInstance: Promise<E.Either<Err, Enforcer>>
export const getEnforcer: EnforceProvider = async () => {
  if (enforcerInstance === undefined) {
    enforcerInstance = createEnforcer()
  }

  return pipe(
    await enforcerInstance,
    E.map(enforcer => enforce(enforcer)),
  )
}

export const enforceRole = (enforceFnProvider: TE.TaskEither<Err, Enforce>) => <
  T extends keyof Actions,
  U extends Actions[T]
>(
  resource: T,
) => (actions: U[]) => (role: Roles): TE.TaskEither<Err, void> =>
  pipe(
    enforceFnProvider,
    TE.chain(enforce => enforce(role, resource, ...actions)),
    TE.chain(result => (result === true ? createVoidTE() : TE.left({ code: 'ROLE_REJECTED' }))),
  )
