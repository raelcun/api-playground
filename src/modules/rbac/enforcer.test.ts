import { taskEither as TE } from 'fp-ts'
import { pipe } from 'fp-ts/lib/pipeable'

import { createEnforcer, getEnforcer } from './enforcer'
import { Roles } from './types'

describe('enforcer', () => {
  test('security policies match snapshot', async () => {
    expect.assertions(1)

    await pipe(
      createEnforcer,
      TE.chain(enforcer => TE.rightTask(() => enforcer.getPolicy())),
      TE.map(policy => expect(policy).toMatchSnapshot()),
    )()
  })

  test('security model matches snapshot', async () => {
    expect.assertions(1)

    await pipe(
      createEnforcer,
      TE.map(enforcer => {
        expect(enforcer.getModel().model).toMatchSnapshot()
      }),
    )()
  })

  test('validation passes for admin with real resource and action', async () => {
    expect.assertions(1)

    const result = await pipe(
      getEnforcer,
      TE.chain(enforce => enforce(Roles.Admin, 'account', 'viewOwn')),
    )()

    expect(result).toMatchSnapshot()
  })

  test('validation passes for admin with fake resource', async () => {
    expect.assertions(1)

    const result = await pipe(
      getEnforcer,
      TE.chain(enforce => enforce(Roles.Admin, 'fakeResource' as any, 'viewAny')),
    )()

    expect(result).toMatchSnapshot()
  })

  test('validation passes for admin with fake action', async () => {
    expect.assertions(1)

    const result = await pipe(
      getEnforcer,
      TE.chain(enforce => enforce(Roles.Admin, 'account', 'fakeAction' as any)),
    )()

    expect(result).toMatchSnapshot()
  })

  test('validation fails for user with insuffient priviledges', async () => {
    expect.assertions(1)

    const result = await pipe(
      getEnforcer,
      TE.chain(enforce => enforce(Roles.User, 'account', 'editAny')),
    )()

    expect(result).toMatchSnapshot()
  })

  test('validation fails for user with fake resource', async () => {
    expect.assertions(1)

    const result = await pipe(
      getEnforcer,
      TE.chain(enforce => enforce(Roles.User, 'fakeResource' as any, 'editAny')),
    )()

    expect(result).toMatchSnapshot()
  })

  test('validation fails for user with fake action', async () => {
    expect.assertions(1)

    const result = await pipe(
      getEnforcer,
      TE.chain(enforce => enforce(Roles.User, 'account', 'fakeAction' as any)),
    )()

    expect(result).toMatchSnapshot()
  })
})
