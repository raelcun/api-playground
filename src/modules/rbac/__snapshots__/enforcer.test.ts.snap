// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enforcer security model matches snapshot 1`] = `
Map {
  "r" => Map {
    "r" => Assertion {
      "key": "r",
      "policy": Array [],
      "rm": DefaultRoleManager {
        "allRoles": Map {},
        "maxHierarchyLevel": 0,
      },
      "tokens": Array [
        "r_sub",
        "r_obj",
        "r_act",
      ],
      "value": "sub, obj, act",
    },
  },
  "p" => Map {
    "p" => Assertion {
      "key": "p",
      "policy": Array [
        Array [
          "user",
          "account",
          "viewOwn",
        ],
        Array [
          "user",
          "account",
          "viewAny",
        ],
      ],
      "rm": DefaultRoleManager {
        "allRoles": Map {},
        "maxHierarchyLevel": 0,
      },
      "tokens": Array [
        "p_sub",
        "p_obj",
        "p_act",
      ],
      "value": "sub, obj, act",
    },
  },
  "e" => Map {
    "e" => Assertion {
      "key": "e",
      "policy": Array [],
      "rm": DefaultRoleManager {
        "allRoles": Map {},
        "maxHierarchyLevel": 0,
      },
      "tokens": Array [],
      "value": "some(where (p_eft == allow))",
    },
  },
  "m" => Map {
    "m" => Assertion {
      "key": "m",
      "policy": Array [],
      "rm": DefaultRoleManager {
        "allRoles": Map {},
        "maxHierarchyLevel": 0,
      },
      "tokens": Array [],
      "value": "g(r_sub, p_sub) && r_obj == p_obj && r_act == p_act || r_sub == \\"admin\\"",
    },
  },
  "g" => Map {
    "g" => Assertion {
      "key": "g",
      "policy": Array [],
      "rm": DefaultRoleManager {
        "allRoles": Map {},
        "maxHierarchyLevel": 0,
      },
      "tokens": Array [],
      "value": "_, _",
    },
  },
}
`;

exports[`enforcer security policies match snapshot 1`] = `
Array [
  Array [
    "user",
    "account",
    "viewOwn",
  ],
  Array [
    "user",
    "account",
    "viewAny",
  ],
]
`;

exports[`enforcer validation fails for user with fake action 1`] = `
Object {
  "_tag": "Right",
  "right": false,
}
`;

exports[`enforcer validation fails for user with fake resource 1`] = `
Object {
  "_tag": "Right",
  "right": false,
}
`;

exports[`enforcer validation fails for user with insuffient priviledges 1`] = `
Object {
  "_tag": "Right",
  "right": false,
}
`;

exports[`enforcer validation passes for admin with fake action 1`] = `
Object {
  "_tag": "Right",
  "right": true,
}
`;

exports[`enforcer validation passes for admin with fake resource 1`] = `
Object {
  "_tag": "Right",
  "right": true,
}
`;

exports[`enforcer validation passes for admin with real resource and action 1`] = `
Object {
  "_tag": "Right",
  "right": true,
}
`;
